legion/engine/llri-dx/source.cpp

```diff
@@ -9,88 +10,143 @@ namespace legion::graphics::llri
+            for (uint32_t i = 0; i < desc.numExtensions; i++)
             {
-                case instance_extension_type::APIValidation:
-                {
-                    extensionCreateResult = internal::createAPIValidationEXT(extension.apiValidation, &output->m_debugAPI);
-                    if (extensionCreateResult == result::Success)
-                        factoryFlags = DXGI_CREATE_FACTORY_DEBUG;
-                    break;
-                }
-                case instance_extension_type::GPUValidation:
+                auto& extension = desc.extensions[i];
+                result extensionCreateResult;
+
+                switch (extension.type)
                 {
-                    extensionCreateResult = internal::createGPUValidationEXT(extension.gpuValidation, &output->m_debugGPU);
-                    break;
+                    case instance_extension_type::APIValidation:
+                    {
+                        extensionCreateResult = internal::createAPIValidationEXT(extension.apiValidation, &output->m_debugAPI);
+                        if (extensionCreateResult == result::Success)
+                            factoryFlags = DXGI_CREATE_FACTORY_DEBUG;
+                        break;
+                    }
+                    case instance_extension_type::GPUValidation:
+                    {
+                        extensionCreateResult = internal::createGPUValidationEXT(extension.gpuValidation, &output->m_debugGPU);
+                        break;
+                    }
+                    default:
+                    {
+                        if (desc.callbackDesc.callback)
+                            desc.callbackDesc(validation_callback_severity::Error, validation_callback_source::Validation, (std::string("createInstance() returned ErrorExtensionNotSupported because the extension type ") + std::to_string((int)extension.type) + " is not recognized.").c_str());
+
+                        extensionCreateResult = result::ErrorExtensionNotSupported;
+                        break;
+                    }
                 }
-                default:
+
+                if (extensionCreateResult != result::Success)
                 {
-                    extensionCreateResult = result::ErrorExtensionNotSupported;
-                    break;
+                    llri::destroyInstance(output);
+                    return extensionCreateResult;
                 }
             }
 
-            if (extensionCreateResult != result::Success)
+            //Store user defined validation callback
+            //DirectX creates validation callbacks upon device creation so we just need to store information about this right now.
+            if (desc.callbackDesc.callback)
+            {
+                output->m_validationCallback = desc.callbackDesc;
+                output->m_validationCallbackMessenger = (void*)enableInternalAPIMessagePolling; //Use as boolean to indicate that a custom callback was set
```

Comment:
Would it be valid to do this then?

```
output->m_validationCallbackMessenger = enableInternalAPIMessagePolling ? (void*)true : (void*)false;
```

Reply:
so `output->m_validationCallbackMessenger` is an integer in this case? cuz in that case i'd say take the bool in as a copy instead of a const reference. a reference, much like a pointer is about the size of an uint64 on 64 bit machines. a bool itself is just 1 byte. so actually copying a bool into the function is cheaper on certain machines than copying a reference or a pointer.

so in that case i'd do:
```cpp
// take enableInternalAPIMessagePolling in by value
output->m_validationCallbackMessenger = reinterpret_cast<void*>(static_cast<std::uintptr_t>(enableInternalAPIMessagePolling));
```

but honestly if you have any `void*` that might not always be pointers, then you should be making them `std::intptr_t` or `std::uintptr_t`. They communicate the usage of possibly being a pointer, or an integer type. casting to and from pointers, and casting to and from integer/boolean types also doesn't invoke undefined behaviour.

Reply:
Makes sense, as mentioned in the other conversation I'll be adjusting this setup because I'm not happy with m_valdiationCallbackMessenger representing both a boolean and a pointer after all